Let CodeClimate decide which engines to run

pointlessone opened the pull request:
This PR removes hardcoded engines and instead lets CC decide which engines to run. This closer matches default CC CLI behaviour.

As a side effect this eliminates a few dependencies as it's not needed to find and parse CC config.

Bonus: fixed end column for issue ranges.

pointlessone committed:
Fix end column of issue range

pointlessone committed:
Run all configured CC engines

Arcanemagus commented on spec/fixtures/README.md:
```diff
@@ -0,0 +1,104 @@
+# One More Time
```
Hilarious, but I don't think we need _this_ big of a file to test that it's working 😉.

Arcanemagus commented on spec/linter-codeclimate-spec.js:
```diff
@@ -13,14 +14,15 @@ describe('The codeclimate provider for Linter', () => {
   it('works with a valid .codeclimate.yml file', () =>
-    waitsForPromise(() =>
+    waitsForPromise(
+      { timeout: 10000 },
```
10 seconds is already the default timeout, no need to specify it.

pointlessone replied:
For me it was failing without this. Reported time was a bit over 5 sec so I assume it's 5 seconds by default. Maybe this changed recently in some of the packages. 

Would you consider a yarn.lock or something else akin to that to lock dependency versions?

Arcanemagus replied:
The 10 second dependency is determined in Atom, since it's what sets up and runs the test environment.

cgalvarez replied:
Atom sets 60 sec on CI, or 5 sec otherwise. This is exactly the same issue that was making the tests fail for me locally. You can read my comment at PR #65.

Arcanemagus commented on spec/linter-codeclimate-spec.js:
```diff
@@ -35,4 +37,21 @@ describe('The codeclimate provider for Linter', () => {
     ),
   );
+
+  it('runs all configured engines', () =>
```
This description is _badly_ misleading. All that this is testing is that it handles a single engine on a single file type...

pointlessone replied:
Agree. The point of this test is to run an engine that was not in the hardcoded ones, and consequently couldn't be ran before. 

I'm open for suggestions both for the spec description and better specs.

Arcanemagus replied:
Since this PR is removing all pre-defined mappings, is this spec actually doing anything different from the one on L22?

pointlessone replied:
Not really. The only difference is that if you run this spec against previous version with hardcoded engines it will fail.

Arcanemagus replied:
Since there no longer are any hardcoded engines in this PR, I'd say just to remove this test since it's doing the same thing as the existing one.

Arcanemagus commented on lib/index.js:
```diff
@@ -192,27 +153,13 @@ export default {
         const execOpts = {
           cwd: dirname(configurationFilePath),
           uniqueKey: `linter-codeclimate::${relpath}`,
+          env: process.env,
```
This is the default, no need to specify it.

pointlessone replied:
It's interesting that you mention this.

For me the the specs were failing the way they fail right now on CI without this line. I hadn't committed it at first thinking that it's just my weird setup (Docker on MacOS runs on non-default ip/port combination for some reason). When I added this line my local setup started picking up `DOCKER_HOST` var and all was well.

That's why when I saw the failure on CI I pushed this line as well. But it appears to be something else this time.

I'd appreciate any hint anyone can give me regarding the CI failure.

Arcanemagus replied:
That's odd, it looks like the only cases where `null` would be returned are if `codeclimate` is crashing or if the process was killed due to the unique spawn constraint.

Since this is failing for you locally can you open the developer tools in the spec window and check the Console? If it's due to `codeclimate` crashing there should be more information printed there.

cgalvarez replied:
I think the CI error does not have to do with the env variables, but the new wrapper for the codeclimate cli. I've been testing it with v0.62.0 and my PR #65 works flawlessly. If I use the recent v0.63.0 or v0.63.1 ones, it fails.

The tests are failing because, as @Arcanemagus says, codeclimate command is failing. If you run the original command `codeclimate analyze -f json -e fixme -e rubocop cool_code.rb </dev/null` you get this output:

```shell
Error: ! analyze
! -f
! json
! -e
! fixme
! -e
! rubocop
! cool_code.rb
! </dev/null
Please supply only one file for analysis
```

If you remove the `</dev/null` argument, then you get this output:

```shell
Error: ! analyze
! -f
! json
! -e
! fixme
! -e
! rubocop
! cool_code.rb
[DEBUG] Is a directory @ io_fillbuf - fd:7 /config.yml
error: (Errno::ENOENT) No such file or directory @ rb_sysopen - /code/.codeclimate.yml
[DEBUG] backtrace: /usr/src/app/lib/cc/analyzer/filesystem.rb:19:in `read'
	/usr/src/app/lib/cc/analyzer/filesystem.rb:19:in `read_path'
	/usr/src/app/lib/cc/cli/analyze.rb:72:in `config'
	/usr/src/app/lib/cc/cli/analyze.rb:76:in `apply_config_options'
	/usr/src/app/lib/cc/cli/analyze.rb:23:in `initialize'
	/usr/src/app/lib/cc/cli/runner.rb:25:in `new'
	/usr/src/app/lib/cc/cli/runner.rb:25:in `run'
	/usr/src/app/lib/cc/cli/runner.rb:9:in `run'
	/usr/src/app/bin/codeclimate:6:in `<main>'
```

I think it would be useful to add the env var `"env":{"CODECLIMATE_DEBUG":1}` if `atom.inDevMode()` to the command options, to attach a more detailed info about the crash to the CI reports and local tests.

The codeclimate command fails, so it returns null, and the tests fail because they expect a JSON object to be returned, but they get `null` instead, and thus cannot access the property '0'.

pointlessone replied:
@cgalvarez Even if the `/dev/null` is the case this PR fails because of something else. I [removed](https://github.com/AtomLinter/linter-codeclimate/pull/67/files#diff-6d186b954a58d5bb740f73d84fe39073R158) it.

cgalvarez replied:
@pointlessone Yes, that's why I've included the output **without** `</dev/null` too, and I can assure that **the tests fail because of the new codeclimate cli wrapper**. It's not an issue of `linter-codeclimate` but `codeclimate cli`.

I you take a look at the CI log, codeclimate is installed at the very beginning of the tests with:

```shell
$ git clone --depth 1 https://github.com/codeclimate/codeclimate.git
$ cd codeclimate && sudo make install && cd ..
```

So it clones the `master` branch of the repo, which is the latest one (v0.63.1). Tests fail with v0.63.0 and v0.63.1. You can clone the repo with the tag `v0.62.0` and you'll see that the tests succeed (and the plugin itself in dev mode).

The `</dev/null` was a comment because since codeclimate cli v0.63.0, it has stopped working, so anyone using this plugin with codeclimate cli v0.63.0+ will think that the problem is related to the plugin, but apparently it is not.

pointlessone replied:
I think I know why it fails with the newer CC CLI.

The primary change in that new CLI can scan piped in source but it needs a filename to determine what engines to run against it. For obvious reasons there can be only one file name specified in that case.

It appears CLI improperly detects this case and errors out because it thinks there are two paths specified.

It thinks there are two paths because Node passes `</dev/null` as a literal parameter to the CLI. It doesn't do redirection as a shell would do. But even if it did I don't see why it's needed here.

In any case, as I mentioned before, this PR removes `</dev/null` bit so the command should not fail because of this specific reason.

@cgalvarez What happens if you directly run `codeclimate analyze -f json -e fixme -e rubocop cool_code.rb` in the terminal?

cgalvarez replied:
When executing `codeclimate analyze -f json -e fixme styles.css` with:

- the current master version of this Atom package (with the `</dev/null` stuff removed)
- and codeclimate v0.63.x
- and a test CSS file

in...

- ... the terminal it works and the command returns the json string:
```
[{"categories":["Bug Risk"],"check_name":"FIXME","description":"FIXME found","location":{"lines":{"begin":20,"end":20},"path":"style.css"},"type":"issue","fingerprint":"b5a5eb86d3be898eebbde98668a6cae5","severity":"minor","engine_name":"fixme"},
{"categories":["Bug Risk"],"check_name":"BUG","description":"BUG found","location":{"lines":{"begin":21,"end":21},"path":"style.css"},"type":"issue","fingerprint":"8e1aa283f29fa773f5cc145033fc0a9c","severity":"minor","engine_name":"fixme"},
{"categories":["Bug Risk"],"check_name":"XXX","description":"XXX found","location":{"lines":{"begin":22,"end":22},"path":"style.css"},"type":"issue","fingerprint":"04678738bf48e8539f9060986d3af186","severity":"minor","engine_name":"fixme"}]
```
- ... atom it fails, I see in the console log the message `linter-codeclimate:: Command: `codeclimate analyze -f json -e fixme style.css` and I get the following error in a notification:
```
Exception details:
- COMMAND: `codeclimate analyze -f json -e fixme style.css`
- MESSAGE: ! analyze
! -f
! json
! -e
! fixme
! style.css
error: (Errno::ENOENT) No such file or directory @ rb_sysopen - /code/.codeclimate.yml
Hide Stack Trace
Error: ! analyze
! -f
! json
! -e
! fixme
! style.css
error: (Errno::ENOENT) No such file or directory @ rb_sysopen - /code/.codeclimate.yml
    at ChildProcess.<anonymous> (/opt/cgalvarez/linter-codeclimate/node_modules/sb-exec/lib/index.js:56:20)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Socket.<anonymous> (internal/child_process.js:334:11)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:493:12)
```

I am **NOT** appending `</dev/null` in any case. This same command works for codeclimate versions prior to v.63.x. That's why I think it is an issue from codeclimate. Maybe it is an issue with the package `sb-exec` used by Helpers, but I'm not into that since I only change the codeclimate version used and it works/fails depending exclusively on that change.

Arcanemagus reviewed (changes requested):
Overall I definitely agree with you on letting the `codeclimate` executable handle as much as possible, I'll leave it to @AtomLinter/linter-codeclimate to comment on whether this is the right change to make though as I can't even run it on my machine to test anything 😛.

Nice catch on the end column! Apparently that's been in there since https://github.com/AtomLinter/linter-codeclimate/commit/9434a6fd20a68aae60e35e2bd573094a3a9d592c 😞.

pointlessone committed:
Add extended timeout for CI

pointlessone commented:
Yesterday Code Climate CLI 0.63.2 was released. It addresses the file name validation.

@cgalvarez You may want to rerun a build on your PR.

@Arcanemagus I addressed all your comments. Could you please take a look at this PR again?

cgalvarez commented:
@pointlessone Thank you very much for the pin.

I've tested with the latest Code Climate (v0.63.5) and the errors are gone :smile: :clap: :clap:

maxjacobson reviewed (approved):
(These changes LGTM)

Arcanemagus commented:
@AtomLinter/linter-codeclimate I don't use this package myself, so others get a higher priority when I have time to work on stuff, so things like this can get somewhat delayed if you leave it up to me 😛.

You should definitely feel free to merge PR's and release updates, if you have any questions on that feel free to @mention me or even better bug me on [Atom's Slack](http://atom-slack.herokuapp.com/).

maxjacobson commented:
Ah right, I forgot that I can merge things :sweat_smile:. Thanks. I definitely have some questions about the build & release process, so I may be bugging you about those on slack :)
